app/scripts/lib/setupSentry.js

```diff
@@ -510,53 +509,70 @@ export default function setupSentry({ release, getState }) {
+   * not been started, start a new sentry session.
    */
   const startSession = async () => {
-    const hub = Sentry.getCurrentHub?.();
-    const options = hub.getClient?.().getOptions?.() ?? {};
-    if (hub && (await getMetaMetricsEnabled()) === true) {
-      options.autoSessionTracking = true;
-      hub.startSession();
+    const isHubDefined = Boolean(Sentry.getCurrentHub());
+    const { isSessionStarted } = global.sentry;
+    const isMetaMetricsEnabled = await getMetaMetricsEnabled();
+
+    if (
+      isHubDefined &&
+      (isSessionStarted === undefined || isSessionStarted === false) &&
+      isMetaMetricsEnabled === true
+    ) {
+      try {
+        Sentry.getCurrentHub().startSession();
+      } finally {
+        Sentry.getCurrentHub().captureSession();
+      }
+
+      global.sentry.isSessionStarted = true;
     }
   };
 
   /**
-   * As long as a reference to the Sentry Hub can be found, and the user has
-   * opted out of MetaMetrics, change the autoSessionTracking option and end
-   * the current sentry session.
+   * As long as a reference to the Sentry Hub can be found, and the session has
+   * already been started, end the current sentry session.
    */
   const endSession = async () => {
-    const hub = Sentry.getCurrentHub?.();
-    const options = hub.getClient?.().getOptions?.() ?? {};
-    if (hub && (await getMetaMetricsEnabled()) === false) {
-      options.autoSessionTracking = false;
-      hub.endSession();
+    const isHubDefined = Boolean(Sentry.getCurrentHub());
+    const { isSessionStarted } = global.sentry;
+    const isMetaMetricsEnabled = await getMetaMetricsEnabled();
+
+    if (
+      isHubDefined &&
+      isSessionStarted === true &&
+      isMetaMetricsEnabled === false
+    ) {
+      try {
+        Sentry.getCurrentHub().captureSession(true);
+      } finally {
+        Sentry.getCurrentHub().endSession();
+      }
+
+      global.sentry.isSessionStarted = false;
     }
   };
 
   /**
-   * Call the appropriate method (either startSession or endSession) depending
-   * on the state of metaMetrics optin and the state of autoSessionTracking on
-   * the Sentry client.
+   * Call the appropriate method (either startSession or endSession). Start a
+   * Sentry session if the `toggleSession` function is being called for the
+   * first time and MetaMetrics was enabled during onboarding, or if a session
+   * has previously been ended. End the Sentry session if a session has
+   * previously been started.
    */
   const toggleSession = async () => {
-    const hub = Sentry.getCurrentHub?.();
-    const options = hub.getClient?.().getOptions?.() ?? {
-      autoSessionTracking: false,
-    };
+    const { isSessionStarted } = global.sentry;
     const isMetaMetricsEnabled = await getMetaMetricsEnabled();
+
     if (
-      isMetaMetricsEnabled === true &&
-      options.autoSessionTracking === false
+      (isSessionStarted === undefined || isSessionStarted === false) &&
```

Comment:
same as https://github.com/MetaMask/metamask-extension/pull/22044/files#r1410623993

Reply:
Previously we had `if (isMetaMetricsEnabled === true) then await startSession();`, which seems correct. Now the code says `if (isMetaMetricsEnabled === false) then await startSession();` which seems incorrect.

If metametrics is not enabled, we don't want any requests going to sentry

Reply:
Confusingly, the logic inside `toggleSession` and inside `startSession` and `endSession` is the opposite. The requests are made appropriately with this logic.

By the time `isMetaMetricsEnabled` is called inside `startSession` or `endSession`, that state has already been changed. See my comment [here](https://github.com/MetaMask/metamask-extension/pull/22044#discussion_r1404507003). I added some console logs to make that more apparent below:
 
**Toggle flow:**
https://github.com/MetaMask/metamask-extension/assets/13814744/91e49a1d-08b2-4c71-9f35-3f8213da312a

**Onboarding flow:**
<img width="1140" alt="Screenshot 2023-11-30 at 14 26 33" src="https://github.com/MetaMask/metamask-extension/assets/13814744/83ccdf30-b61e-4373-890e-283b490d1897">